torchrl/objectives/cql.py

```diff
@@ -0,0 +1,582 @@
+)
+from torchrl.objectives.value import TD0Estimator, TD1Estimator, TDLambdaEstimator
+
+try:
+    try:
+        from torch import vmap
+    except ImportError:
+        from functorch import vmap
+
+    _has_functorch = True
+    err = ""
+except ImportError as err:
+    _has_functorch = False
+    FUNCTORCH_ERROR = err
+
+
+class CQLLoss(LossModule):
+    """TorchRL implementation of the CQL loss.
+
+    Presented in "Conservative Q-Learning for Offline Reinforcement Learning" https://arxiv.org/abs/2006.04779
+
+    Args:
+        actor_network (ProbabilisticActor): stochastic actor
+        qvalue_network (TensorDictModule): Q(s, a) parametric model.
+            This module typically outputs a ``"state_action_value"`` entry.
+
+        loss_function (str, optional): loss function to be used with
+            the value function loss. Default is `"smooth_l1"`.
+        alpha_init (float, optional): initial entropy multiplier.
+            Default is 1.0.
+        min_alpha (float, optional): min value of alpha.
+            Default is None (no minimum value).
+        max_alpha (float, optional): max value of alpha.
+            Default is None (no maximum value).
+        fixed_alpha (bool, optional): if ``True``, alpha will be fixed to its
+            initial value. Otherwise, alpha will be optimized to
+            match the 'target_entropy' value.
+            Default is ``False``.
+        target_entropy (float or str, optional): Target entropy for the
+            stochastic policy. Default is "auto", where target entropy is
+            computed as :obj:`-prod(n_actions)`.
+        delay_actor (bool, optional): Whether to separate the target actor
+            networks from the actor networks used for data collection.
+            Default is ``False``.
+        delay_qvalue (bool, optional): Whether to separate the target Q value
+            networks from the Q value networks used for data collection.
+            Default is ``True``.
+        gamma (float, optional): Discount factor. Default is ``None``.
+        temperature (float, optional): CQL temperature. Default is ``1.0``.
+        min_q_weight (float, optional): Minimum Q weight. Default is ``1.0``.
+        max_q_backup (bool, optional): Whether to use the max-min Q backup.
+            Default is ``False``.
+        deterministic_backup (bool, optional): Whether to use the deterministic. Default is ``True``.
+        num_random (int, optional): Number of random actions to sample for the CQL loss.
+            Default is ``10``.
+        with_lagrange (bool, optional): Whether to use the Lagrange multiplier.
+            Default is ``False``.
+        lagrange_thresh (float, optional): Lagrange threshold. Default is ``0.0``.
+        priority_key (str, optional): [Deprecated, use .set_keys(priority_key=priority_key) instead]
+            Tensordict key where to write the
+            priority (for prioritized replay buffer usage). Defaults to ``"td_error"``.
```

Comment:
Could you add a simple example?
Just copy paste the networks and loss construction from the tests, it should be sufficient.